### DISCUSSION TITLE
How to get instance's full before and after state in the `before_update` event?

### DISCUSSION THREAD

**Discussion by seperman** [sqlalchemy/sqlalchemy#8435] [Usage Questions] [answered]
Hello,
Is there a way to get an instance's full before and after states in the `before_update` event?
The current method I'm using:

```py
def before_update_listener_for_history(mapper, connection, obj):
    """
    This listener will update the history attribute of any model that it is attached to via events.
    """
    state = inspect(obj)
    if state.committed_state:
        changes = {}
        for attr in state.attrs:
            hist = attr.load_history()
```

Problem:

It doesn't have the "before" state for json data type.

Ideal solution:

I want to just have a `dict` of all the attributes of an instance before there were any changes in the session. The current state is already accessible via `obj.__dict__`. Then I can get the delta of the 2 states myself.

Thanks

**Comment by zzzeek** [comment]
hi -

I have to make some guesses what you are doing, are you using JSON with [Mutable](https://docs.sqlalchemy.org/en/14/orm/extensions/mutable.html?highlight=mutable#module-sqlalchemy.ext.mutable) and making changes in place?    an in-place change doesn't make a copy of the object.   you would have to subclass the colleciton, like [MutableDict](https://docs.sqlalchemy.org/en/14/orm/extensions/mutable.html?highlight=mutable#sqlalchemy.ext.mutable.MutableDict) , and when changes are intercepted, make a copy of the whole structure somewhere.

**Comment by seperman** [comment, marked as answer]
Hi @zzzeek 
Thanks for the prompt response. Ok I am indeed using the MutableDict. Are you saying I should add a new field to the object to store the previous state as a json in the database? What I was hoping was that somewhere SQLalchemy is holding on the the "before" state of the object so when I do changes in place in the MutableDict, I can get the "before" state. It is doing it already for all the other data types from what I see in the attr.history

**Reply by zzzeek** [reply]
yes that's what it does for regular scalar values when you say:

```
   x.foo = 5
   session.flush()    # flushes '5'
   # later
   x.foo = 10    # we replace the 5 with 10, we have both
```

But this doesnt' apply to "mutable".  we would have to copy the entire dictionary every time you mutate any value within it which would be extremely inefficient.
   
I think what you need to do is either find a way to solve the problem you have in a different way, or otherwise you need to code an explicit step to take a snapshot of that JSON structure before its mutated.   Subclassing MutableDict and adding more logic to the [changed()](https://github.com/sqlalchemy/sqlalchemy/blob/5c7ab1f808b20fe5e58d83a738b201a7eaa8973f/lib/sqlalchemy/ext/mutable.py#L538) to do what you need is one way to achieve this.

**Reply by seperman** [reply]
Ok Awesome. That seems to be working. Thank you!
